### DISCUSSION TITLE
Custom python package installation

### DISCUSSION THREAD

**Discussion by sbhonagiri-evoke** [langgenius/dify#4393] [Suggestion]
### Self Checks

- [X] I have searched for existing issues [search for existing issues](https://github.com/langgenius/dify/issues), including closed ones.
- [X] I confirm that I am using English to submit this report (我已阅读并同意 [Language Policy](https://github.com/langgenius/dify/issues/1542)).
- [X] Please do not modify this template :) and fill in all the required fields.

### 1. Is this request related to a challenge you're experiencing?

i have to run a separate python endpoint since dify doesn't has the ability

### 2. Describe the feature you'd like to see

i just need an option to enter the pip package name and it should be installed in the backend to use in code node

### 3. How will this feature improve your workflow or experience?

this will improve the overall user experience and reduce lot of manual work

### 5. Can you help us with this feature?

- [X] I am interested in contributing to this feature.

**Comment by huy302** [comment]
I'm interested in this feature as well, Dify would be so much more useful if we are allowed to install additional python packages to run in the code block

**Comment by EchterTimo** [comment]
@sbhonagiri-evoke 
@huy302


Little late but I found what you need:

> [!WARNING]
> This only works if you are on a selfhosted instance. You can not add custom packages if you are using [cloud.dify.ai](https://cloud.dify.ai)

In my case I am on a docker based deployment: https://docs.dify.ai/en/self-host/quick-start/docker-compose

## Steps

1. create file `/docker/dependencies/python-requirements.txt`
2. Enter Python libs in that file like you do with requirements.txt. Versioning your packages is advised if you don't want new releases breaking your code.

```pip-requirements
dify-client==0.1.10
```

> [!WARNING]
> Libs like numpy need extra SYSCALLS permissions that you need to specify. This is mentioned by the lead maintainer in this discussion: https://github.com/langgenius/dify/discussions/8200#discussioncomment-10598903
> However if you are just using an api wrapper like [dify-client](https://github.com/langgenius/dify-python-sdk) you should be fine without extra permissions.
> Feel free to try other libs and share your experience.

>[!WARNING]
> The default Sanbox Python version is 3.10.17.
> Make sure the libs you want to use are compatible with that version.
> You can check the version by using a code node with the following code.


```python
import sys
def main():
    return {
        "version": sys.version
    }
```


>[!NOTE]
> Here you can find out what dependencies the dify sandbox already installs for you:
> https://github.com/langgenius/dify-sandbox/blob/3f563d94d9ba2201c495f5653e2830745331c478/docker/templates/production.dockerfile#L4
> 
> Currently these are: `httpx`, `requests`, `jinja2`, `PySocks`, `httpx[socks]`

3. Finally run this command to restart the sandbox

```bash
sudo docker restart docker-sandbox-1
```

4. Verify the container is running by using this command:
```bash
docker ps -a | grep -e sandbox -e "CONTAINER ID"
```

5. Check the logs of the sandbox container to see if the dependencies were installed correctly. You can use this command to see the last 300 lines of logs:

```bash
docker compose logs --tail=300 sandbox
```

You want to have the following line in the logs:
```log
sandbox-1  | 2026/02/19 20:56:51 server.go:60: [INFO]python dependencies sandbox initialized
```

6. You should now be able to use the libs in the sandbox

## Where do I have this information from?

- https://github.com/langgenius/dify/discussions/8200
- https://forum.dify.ai/t/how-can-i-install-python-packages-in-the-code-node/55
- https://github.com/langgenius/dify-sandbox/blob/main/FAQ.md
- testing it out myself
